### PR TITLE
build: Add --enable-c++20 option

### PR DESCRIPTION
**Pull request by maflcko**
This is for CI and devs only and doesn't change that C++17 is the standard we are currently using. The option `--enable-c++20` allows CI to check that the C++17 code in the repo is  also valid C++20. (There are some cases where valid C++17 doesn't compile under C++20).

Also, it allows developers to easily play with C++20 in the codebase.

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
You may or may not be interested in one of my PRs to dash that did this https://github.com/dashpay/dash/pull/4600

I was planning on doing the equivalent here, but didn't want to until we could merge https://github.com/bitcoin/bitcoin/pull/23340 which if I recall correctly is required to fix cxx20 builds

**Comment by fanquake** [comment]
Support for c++20 has now been merged into the `ax_cxx_compile_stdcxx.m4` macro upstream: https://github.com/autoconf-archive/autoconf-archive/pull/244.

**Comment by MarcoFalke** [comment]
Looks like this was broken after rebasing on the std::fs changes. Marking as draft again

**Comment by fanquake** [comment]
If you want to split the macro update off, happy to merge that earlier, given it's a mechanical change.

**Review by felipsoarez — COMMENTED** [review]
utACK

**Comment by MarcoFalke** [comment]
This is still waiting on #24404, so please review that first.

**Comment by MarcoFalke** [comment]
Rebased, Taken out of Draft

**Comment by fanquake** [comment]
Concept ACK. I think having the developer option, and carrying some number of code changes to support c++20 compilation is fine, but it's going to be a long time before we can make it a requirement.

**Comment by MarcoFalke** [comment]
Force pushed to address feedback: https://github.com/bitcoin/bitcoin/pull/24531#discussion_r828566749

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa442133f1ea37edb7c769590da9592b541fce6c. All these changes look safe to me, and it is definitely better if the code is tested with c++20 and compiles without changes.

A lot of changes are made in the PR without an explanation about the reason, though. Like the enum-enum warning change and scheduler lambda change. Ideally the changes would be broken up more, with one change per commit, and the commit messages would explain the reasons behind the changes, or at least give hints.

**Comment by MarcoFalke** [comment]
Thx, done. (No code changes)

**Review by ryanofsky — APPROVED** [review]
Code review ACK fab63ca14298d37022e23d3e9747bc4bf94121c9. Thanks for the descriptions, I understand this much better now!

**Comment by fanquake** [comment]
@theuni you might also be interested here

**Review by theStack — APPROVED** [review]
Concept and code-review ACK fab63ca14298d37022e23d3e9747bc4bf94121c9 :two: :zero:

Tested `--enable-c++20` compilation with clang 11.1.0 (OpenBSD 7.0).

**Comment by hebasto** [comment]
Concept ACK.

**Review by hebasto — COMMENTED** [review]
Approach ACK fab63ca14298d37022e23d3e9747bc4bf94121c9, except for the fa08f8b8cd4eb13cc5904d33e027f6c485b9f4b1 "Set -Wno-deprecated-enum-enum-conversion" commit. An alternative has been suggested in #24624.

**Comment by MarcoFalke** [comment]
I'd say the two pull requests are mostly orthogonal. If one is merged before the other, the one merged second can drop the temporary `-Wno` commit.

**Review by hebasto — APPROVED** [review]
ACK fab63ca14298d37022e23d3e9747bc4bf94121c9

> I'd say the two pull requests are mostly orthogonal. If one is merged before the other, the one merged second can drop the temporary `-Wno` commit.

Agree.

**Comment by MarcoFalke** [comment]
Removed a commit. Should be trivial to re-ACK.

**Review by hebasto — APPROVED** [review]
~re-ACK fa6bc3b2ccb7386ed496dd8be6b748e95c28b957~

CI failure?

**Comment by MarcoFalke** [comment]
Please review #24641 first, while this fails to compile.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 999982b06ce1d1280e5ce48f9253c6c536f41a12. Since last review was rebased, and enum-conversion change was dropped, and CSerializedNetMsg copy workaround was added

**Review by fanquake — APPROVED** [review]
utACK 999982b06ce1d1280e5ce48f9253c6c536f41a12

**Comment by Sjors** [comment]
Nice! Though macOS trips over a deprecation (warning), see #24682.

**Pull request merged**
